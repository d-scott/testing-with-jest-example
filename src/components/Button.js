import React from 'react';
import PropTypes from 'prop-types';

class Button extends React.Component {
    static get defaultProps() {
        return {
            enabled: true,
            visible: true,
            label: '',
            icon: ''
        };
    }

    constructor(props) {
        super(props);

        this._onClick = this._onClick.bind(this);
    }

    render() {
        const className = this._getClassName();

        return (
            <button className={ className } onClick={ this._onClick }>
                { this.props.label }
            </button>
        );
    }

    _onClick(e) {
        if (this.props.enabled) {
            this.props.onClick(e);
        }
    }

    _getClassName() {
        let className = this.props.className || 'button';

        if (!this.props.enabled) {
            className = `${className} disabled`;
        }

        if (!this.props.visible) {
            className = `${className} hidden`;
        }

        return className;
    }
}

Button.propTypes = {
    onClick: PropTypes.func,
    enabled: PropTypes.bool,
    visible: PropTypes.bool,
    label: PropTypes.string,
    icon: PropTypes.node
};

export default Button;
