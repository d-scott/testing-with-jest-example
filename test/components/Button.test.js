import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import Button from '../../src/components/Button';

describe('----Button----', () => {

    test('should render a .button', () => {
        const component = renderer.create(
            <Button />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render a .button with a label', () => {
        const component = renderer.create(
            <Button label='MyButton' />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render a disabled .button', () => {
        const component = renderer.create(
            <Button enabled={false} />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render a hidden .button', () => {
        const component = renderer.create(
            <Button visible={false} />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should have default props', () => {
        const wrapper = mount(<Button/>);
        expect(wrapper.props()).toEqual(expect.objectContaining({
            enabled: true,
            visible: true,
            label: '',
            icon: ''
        }));
    });

    test('should not call callback if disabled', () => {
        const onClick = jest.fn();
        const wrapper = mount(<Button onClick={onClick} />);
        wrapper.find('button').simulate('click');
        expect(onClick).toHaveBeenCalled();

        wrapper.setProps({ enabled: false });
        wrapper.find('button').simulate('click');
        expect(onClick).toHaveBeenCalledTimes(1);
    });

});
