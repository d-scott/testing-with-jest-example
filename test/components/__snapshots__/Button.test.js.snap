// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`----Button---- should render a .button 1`] = `
<button
  className="button"
  onClick={[Function]}
>
  
</button>
`;

exports[`----Button---- should render a .button with a label 1`] = `
<button
  className="button"
  onClick={[Function]}
>
  MyButton
</button>
`;

exports[`----Button---- should render a disabled .button 1`] = `
<button
  className="button disabled"
  onClick={[Function]}
>
  
</button>
`;

exports[`----Button---- should render a hidden .button 1`] = `
<button
  className="button hidden"
  onClick={[Function]}
>
  
</button>
`;
